---
# TODO: keep inline r code and params$title only if this will be set dynamically
title: '`r params$title`'
author: 'University of Melbourne Centre for Cancer Research'
date: '`r Sys.time()`'
output:
  html_document:
    theme: cosmo
    toc: true
    toc_float: true
code_download: true
rmdformats::material:
  highlight: kate
params:
  title: 'File Comparison Report'
  results_directory: '~/projects/pipeline_comparison/output/'
---

```{r knitr_options, include=FALSE}
knitr::opts_chunk$set(
  collapse=TRUE,
  echo=FALSE,
  warning=FALSE,
  message=FALSE
)
```

```{r libraries}
library(bedr)
library(DT)
library(tidyverse)
```

```{r sample_input_directories}
v.files_exclude <- c(
  'report.html',
  'input_files.tsv'
)
v.results_directory_children <- Sys.glob(paste0(params$results_directory, '/*'))
v.sample_dirs <- v.results_directory_children[! basename(v.results_directory_children) %in% v.files_exclude]
```

# __Overview__
```{r overview, message=TRUE}
message('TODO: A visual summary and high-level statistics')
```

# __Small variants__
## __Summary__
```{r smlv_summary_constants}
# Small variant summary table columns
# Order and display name
v.smlv_summary_table_columns <- c(
  'Sample name'='sample',
  'VCF type'='flabel',
  'Subset'='subset',
  'Run 1 count'='run1_count',
  'Run 2 count'='run2_count',
  'SNP recall'='SNP_Recall',
  'SNP precision'='SNP_Precision',
  'INDEL recall'='IND_Recall',
  'INDEL precision'='IND_Precision',
  'SNP truth'='SNP_Truth',
  'SNP TP'='SNP_TP',
  'SNP FP'='SNP_FP',
  'SNP FN'='SNP_FN',
  'INDEL truth'='IND_Truth',
  'INDEL TP'='IND_TP',
  'INDEL FP'='IND_FP',
  'IND FN'='IND_FN'
)
# Rounded columns
v.smlv_summary_table_columns_round <- c(
  'SNP recall',
  'SNP precision',
  'INDEL recall',
  'INDEL precision'
)
# Columns formatted with commas
v.smlv_summary_table_columns_comma <- c(
  'Run 1 count',
  'Run 2 count',
  'SNP truth',
  'SNP TP',
  'SNP FP',
  'SNP FN',
  'INDEL truth',
  'INDEL TP',
  'INDEL FP',
  'IND FN'
)

s.bgsize <- '90% 90%'
```

```{r smlv_summary_data}
get_smlv_summary_data <- function(s.base_dir) {
  # Read in counts
  d.counts <- readr::read_tsv(paste0(s.base_dir, '/small_variants/counts.tsv'), col_types=readr::cols())
  # Select and read in all comparison files
  v.comp_fps <- Sys.glob(paste0(s.base_dir, '/small_variants/3_comparison/*tsv'))
  v.comp <- lapply(v.comp_fps, readr::read_tsv, col_types=readr::cols())

  # Process counts data
  # 1. remove intersect VCFS - identified by run == 'n/a'
  # 2. recode values: source: input -> all; run: first -> run1_count, second -> run2_count
  # 3. cast to wider format - remove filename col; add first and second columns
  d.counts_wide <- d.counts %>%
    dplyr::filter(run!='n/a') %>%
    dplyr::mutate(
      source=dplyr::recode(source, input='all'),
      run=dplyr::recode(run, first='run1_count', second='run2_count'),
    ) %>%
    tidyr::pivot_wider(names_from=run, id_cols=c(vcf_type, source), values_from=count)

  # Create small variants table
  # 1. bind individual comparison tables into one
  # 2. full join on comparison ~ counts; i.e. add counts to comparison table
  # 3. select, order, and rename desired columns
  # 4. round precision/recall and add commas to other columns
  d.smlv_table <- dplyr::bind_rows(v.comp) %>%
    dplyr::full_join(d.counts_wide, by=c('flabel'='vcf_type', 'subset'='source')) %>%
    dplyr::select(dplyr::all_of(v.smlv_summary_table_columns)) %>%
    dplyr::mutate(dplyr::across(dplyr::all_of(v.smlv_summary_table_columns_round), function(v.col) { format(round(v.col, 4), nsmall=4) })) %>%
    dplyr::mutate(dplyr::across(dplyr::all_of(v.smlv_summary_table_columns_comma), scales::comma))
  return(d.smlv_table)
}

# Finalise small variant data
# 1. bind rows into single table
# 2. sort rows by subset
d.smlv_summary_data <- purrr::map(v.sample_dirs, get_smlv_summary_data) %>%
  dplyr::bind_rows() %>%
  dplyr::arrange(Subset)
```

```{r smlv_summary_table}
d.smlv_summary_data %>%
  DT::datatable(
    rownames=FALSE,
    filter=list(position='top', clear=FALSE, plain=TRUE),
    extensions='Scroller',
    options=list(scroller=TRUE, scrollX=TRUE, scrollY=600)
  ) %>%
  DT::formatStyle(
    v.smlv_summary_table_columns_round,
    background=DT::styleColorBar(as.numeric(unlist(d.smlv_summary_data[ ,v.smlv_summary_table_columns_round])), 'lightgreen'),
    backgroundSize=s.bgsize,
    backgroundRepeat='no-repeat',
    backgroundPosition='center'
  )
```

## __PCGR tables__
```{r smlv_pcgr_constants}
# Small variant PCGR table columns
# Order and display name
v.smlv_vcf_columns <- c(
  'Sample name'='sample_name',
  'Chromosome'='CHROM',
  'Position'='POS',
  'PCGR tier'='PCGR_TIER',
  'PCGR symbol'='PCGR_SYMBOL',
  'Quality score'='QUAL',
  'Tumor AF'='TUMOR_AF',
  'Normal AF'='NORMAL_AF',
  'Callers'='CALLERS',
  'SAGE hotspot'='SAGE_HOTSPOT'
)
# bcftools isec filenames and call types
v.files <- c(
  'fp'= '0000.vcf',
  'fn'='0001.vcf'
)
# PCGR variant type display order and names
v.pcgr_variant_types <- c(
  'Tier 1'='TIER_1',
  'Tier 2'='TIER_2',
  'Tier 3'='TIER_3',
  'Tier 4'='TIER_4',
  'Non-coding'='NONCODING'
)
```

```{r smlv_pcgr_data}
get_smlv_pcgr_data <- function(s.base_dir) {
  s.sample_name <- basename(s.base_dir)
  v.pcgr_vcfs <- purrr::map(names(v.files), function(s.vcf_type) {
    # Collect variant data
    # 1. read in vcf, select data section, and convert to tibble
    # 2. select specific rows, if they are present (using `values` to delay renaming columns until later)
    # 3. add column for (1) sample name, and (2) to identify variant as fp or fn
    # 4. add missing columns
    s.vcf_filename <- v.files[s.vcf_type]
    d.vcf <- bedr::read.vcf(paste0(s.base_dir, '/small_variants/2_variants_intersect/input/pcgr/', s.vcf_filename), split.info=TRUE, verbose=FALSE) %>%
      purrr::pluck('vcf') %>%
      tibble::as_tibble() %>%
      dplyr::select(unname(v.smlv_vcf_columns)[v.smlv_vcf_columns %in% colnames(.)]) %>%
      dplyr::mutate(sample_name=s.sample_name, call_errortype=s.vcf_type)
    # Add any missing columns
    v.columns_missing <- v.smlv_vcf_columns[! v.smlv_vcf_columns %in% colnames(d.vcf)]
    v.columns_missing <- v.columns_missing[v.columns_missing!='sample_name']
    d.vcf[ ,v.columns_missing] <- NA
    # Rename and order columns
    d.vcf <- dplyr::select(d.vcf, dplyr::all_of(v.smlv_vcf_columns))
    # Return NULL if vcf is empty
    if (nrow(d.vcf)==0) {
      return(NULL)
    } else {
      return(d.vcf)
    }
  })
  # If both vcfs contain no data, return NULL
  if (all(sapply(v.pcgr_vcfs, is.null))) {
    return(NULL)
  }
  # Otherwise, finalise variant data
  # 1. merge fp and fn into single tibble
  # 2. order entries by chromosome number
  # 3. place the 'sample_name' column first
  # 4. format position with commas
  dplyr::bind_rows(v.pcgr_vcfs) %>%
    dplyr::arrange(Chromosome) %>%
    dplyr::select('Sample name', dplyr::everything()) %>%
    dplyr::mutate(Position=scales::comma(Position))
}

# Finalise PCGR data
# 1. bind rows into single table
# 2. sort rows by subset
d.smlv_pcgr_data <- purrr::map(v.sample_dirs, get_smlv_pcgr_data) %>%
  dplyr::bind_rows()
```

```{r smlv_pcgr_tablulate_function}
# Set missing PCGR categories; required for selecting and ordering columns
v.missing <- v.pcgr_variant_types[! v.pcgr_variant_types %in% unique(d.smlv_pcgr_data$`PCGR tier`)]

tabulate_pcgr_tiers <- function(s.var) {
  # Create tabulated counts
  # 1. group data by PCGR tier and given variable
  # 2. tally counts
  # 3. convert from grouped tbl back to ungrouped
  # 3. cast into wide format
  d.table_counts <- d.smlv_pcgr_data %>%
    dplyr::group_by(!!dplyr::sym(s.var), `PCGR tier`) %>%
    dplyr::count(.drop=FALSE) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_wider(names_from=`PCGR tier`, id_cols=!!dplyr::sym(s.var), values_from=n, values_fill=0)
  # Complete processing of tabulated data
  # 1. add missing rows
  # 2. order and rename with grouping variable in first position and unknown categories at end
  # 3. add botton sum margin
  # 4. set rowwise, and then compute and add right sum margin
  # 5. convert from rowwise tbl to ungrouped
  d.table_counts[ ,v.missing] <- 0
  dplyr::select(d.table_counts, c(!!dplyr::sym(s.var), dplyr::all_of(v.pcgr_variant_types)), dplyr::everything()) %>%
    dplyr::bind_rows(
      dplyr::summarise(
        .,
        dplyr::across(where(is.numeric), sum),
        dplyr::across(where(is.character), ~'Total'))
    ) %>%
    dplyr::rowwise(.) %>%
    dplyr::mutate(Total=sum(dplyr::c_across(where(is.numeric)))) %>%
    dplyr::ungroup()
}
```

### __PCGR__ variants counts by sample
```{r smlv_pcgr_table_sample_name}
# Tabulate by sample name
tabulate_pcgr_tiers('Sample name') %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width=FALSE, position='left')
```

### __PCGR__ variants counts by SAGE hotspot
```{r smlv_pcgr_table_sage}
# Tabulate by sage hotspot
tabulate_pcgr_tiers('SAGE hotspot') %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width=FALSE, position='left')
```

### __PCGR__ variants
```{r smlv_pcgr_table_all_variants}
# All PCGR variants
d.smlv_pcgr_data %>%
  DT::datatable(
    rownames=FALSE,
    caption='PCGR FP and FN variants for all samples.',
    filter=list(position='top',clear=FALSE, plain=TRUE),
    extensions=c('Scroller', 'Buttons'),
    options=list(scroller=TRUE, scrollX=TRUE, scrollY=220, dom='Bfrtip', buttons=c('csv'))
  )
```

# __Structural variants__
```{r}
v.sv_summary_table_columns <- c(
  'Sample name'='sample',
  'VCF type'='flabel',
  'Run 1 count'='run1_count',
  'Run 2 count'='run2_count',
  'Recall'='Recall',
  'Precision'='Precision',
  'Truth'='Truth',
  'TP'='TP',
  'FP'='FP',
  'FN'='FN'
)
v.sv_summary_table_columns_big_mark <- c(
  'Run 1 count',
  'Run 2 count',
  'Truth',
  'TP',
  'FP',
  'FN'
)

v.sv_fpfn_table_columns <- c(
  'Call error'='FP_or_FN',
  'Sample name'='sample',
  'VCF type'='flabel',
  'Chromosome 1'='chrom1',
  'Position 1'='pos1',
  'Chromosome 2'='chrom2',
  'Position 2'='pos2',
  'SV type'='svtype'
)
v.sv_fpfn_table_columns_big_mark <- c(
  'Position 1',
  'Position 2'
)
```

## __Summary__
```{r}
d.sv_summary_data <- purrr::map(v.sample_dirs, function(s.base_dir) {
  readr::read_tsv(
    paste0(s.base_dir, '/structural_variants/eval_metrics.tsv'),
    col_types=readr::cols()
  )
}) %>%
  dplyr::bind_rows() %>%
  dplyr::select(all_of(v.sv_summary_table_columns))
```

```{r}
sv_tpar <- list(
  # scrollY should be 35px per row, with max of 580px.
  scroll_y=10 + min(nrow(d.sv_summary_data) * 35, 570),
  bgsize='90% 90%',
  big_mark_cols=v.sv_summary_table_columns_big_mark,
  bar_colour='#99ff99',
  bar_range=c(0.97, 1)
)

d.sv_summary_data %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames=FALSE,
    #caption=sv_cap,
    class='cell-border display compact',
    filter=list(position='top', clear=FALSE, plain=FALSE),
    extensions=c('Scroller', 'KeyTable'),
    options=list(scroller=TRUE, scrollX=TRUE, scrollY=sv_tpar$scroll_y)) %>%
  DT::formatStyle(
    c('Recall', 'Precision'),
    background=DT::styleColorBar(sv_tpar$bar_range, color=sv_tpar$bar_colour),
    backgroundSize=sv_tpar$bgsize, backgroundRepeat='no-repeat', backgroundPosition='center'
  ) %>%
  DT::formatCurrency(sv_tpar$big_mark_cols, currency='', interval=3, mark=',', digits=0)
```
## __Call error counts__
```{r}
# FP/FN table
d.sv_fpfn_data <- purrr::map(v.sample_dirs, function(s.base_dir) {
  d <- readr::read_tsv(
    paste0(s.base_dir, '/structural_variants/fpfn.tsv'),
    col_types='ccccicic'
  )
  if (nrow(d)==0) {
    return(NULL)
  } else {
    return(d)
  }
}) %>%
  dplyr::bind_rows() %>%
  dplyr::select(all_of(v.sv_fpfn_table_columns))
```

```{r}
sv_fpfn_tpar <- list(
  # scrollY should be 35px per row, with max of 580px.
  scroll_y=10 + min(nrow(d.sv_fpfn_data) * 35, 570),
  bgsize='90% 90%',
  big_mark_cols=v.sv_fpfn_table_columns_big_mark,
  bar_colour='#99ff99',
  bar_range=c(0.97, 1))

d.sv_fpfn_data %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames=FALSE,
    class='cell-border display compact',
    filter=list(position='top', clear=FALSE, plain=FALSE),
    extensions=c('Scroller', 'KeyTable'),
    options=list(
      scroller=TRUE,
      scrollX=TRUE,
      scrollY=sv_fpfn_tpar$scroll_y)
  ) %>%
  DT::formatStyle(
    'Call error',
    backgroundColor=DT::styleEqual(
      c('fp', 'fn'),
      c('#99ff99', '#FFCCCB')
    )
  ) %>%
  DT::formatCurrency(
    sv_fpfn_tpar$big_mark_cols,
    currency='',
    interval=3,
    mark=',',
    digits=0
  )
```

## __SV circos__ {.tabset .tabset-pills}
```{r results='asis'}
d.sv_circos_data <- purrr::map(v.sample_dirs, function(s.base_dir) {
  sample_name <- basename(s.base_dir)
  tibble::tibble(
    'Sample name'=sample_name,
    png_filepath=Sys.glob(paste0(s.base_dir, '/structural_variants/circos/circos_*.png'))
  )
}) %>%
  dplyr::bind_rows() %>%
  dplyr::left_join(d.sv_summary_data, by=c('Sample name'))

for (i in 1:nrow(d.sv_circos_data)) {
  cat('\n### ', paste(d.sv_circos_data$`Sample name`[i], '\n'))
  cat(glue::glue('<img src="{d.sv_circos_data$png_filepath[i]}" alt="{d.sv_circos_data$`Sample name`[i]}" width="30%">'), '\n')
  cat('\n***\n')

}
```

# __Copy number variants__
## __Summary__
```{r}
s.umcccr_genes_fp <- system.file('extdata/genes/umccr_cancer_genes.latest.genes', package='woofr')
v.umccr_genes <- readr::read_lines(s.umcccr_genes_fp)

v.cnv_table_columns <- c(
  'Sample name'='sample_name',
  'Gene'='gene',
  'UMCCR gene'='in_umccr',
  'Min different'='min_diff',
  'Max different'='max_diff',
  'Min (first/second)'='mincn_1_2',
  'Max (first/second)'='maxcn_1_2',
  'Min diff'='mincn_diff',
  'Max diff'='maxcn_diff',
  'Chromosome'='chrom',
  'Start'='start',
  'End'='end'
)
v.cnv_table_columns_comma <- c(
  'Start',
  'End'
)
v.cnv_table_columns_colour <- c(
  'UMCCR gene',
  'Min different',
  'Max different'
)

# Prepare copy number variant data
# 1. read individual CN data files, and add sample name column
# 2. bind rows together, then create new columns:
#      - combine min/max for run 1 and run 2,
#      - difference of min/max b/n runs, and
#      - is a UMCCR gene
# 3. order and rename colums
# 4. add format appropriate numeric columns with commas
d.cnv_data_diff <- purrr::map(v.sample_dirs, function(s.base_dir) {
  s.sample_name <- basename(s.base_dir)
  readr::read_tsv(paste0(s.base_dir, '/copy_number_variants/cn_diff.tsv'), col_types='ciicddddcc') %>%
    dplyr::mutate(sample_name=s.sample_name)
}) %>%
  dplyr::bind_rows() %>%
  dplyr::mutate(
    mincn_1_2=paste0(min_cn.run1, '/', min_cn.run2),
    maxcn_1_2=paste0(max_cn.run1, '/', max_cn.run2),
    mincn_diff=round(abs(min_cn.run1 - min_cn.run2), 1),
    maxcn_diff=round(abs(max_cn.run1 - max_cn.run2), 1),
    in_umccr=ifelse(gene %in% v.umccr_genes, 'TRUE', 'FALSE')
  ) %>%
  dplyr::select(dplyr::all_of(v.cnv_table_columns)) %>%
  dplyr::mutate(across(dplyr::all_of(v.cnv_table_columns_comma), scales::comma))
```

```{r}
d.cnv_data_diff %>%
  dplyr::mutate_if(is.character, as.factor) %>%
  DT::datatable(
    rownames=FALSE,
    class='cell-border display compact',
    caption=paste0('Differences in Copy Number (threshold: 0.1)'),
    filter=list(position='top', clear=FALSE, plain=FALSE),
    extensions=c('Scroller', 'KeyTable'),
    options=list(
      scroller=TRUE,
      scrollX=TRUE,
      scrollY=10 + min(nrow(d.cnv_data_diff) * 35, 570)
    )
  ) %>%
  DT::formatStyle(
    v.cnv_table_columns_colour,
    backgroundColor=DT::styleEqual(c('TRUE'), c('#FFCCCB'))
  )
```
