# Small variants
## Summary
```{r smlv_summary_constants}
# Small variant summary table columns
# Order and display name
v.smlv_summary_table_columns <- c(
  'Sample name'='sample',
  'VCF type'='flabel',
  'Subset'='subset',
  'Run 1 count'='run1_count',
  'Run 2 count'='run2_count',
  'SNP recall'='SNP_Recall',
  'SNP precision'='SNP_Precision',
  'INDEL recall'='IND_Recall',
  'INDEL precision'='IND_Precision',
  'SNP truth'='SNP_Truth',
  'SNP TP'='SNP_TP',
  'SNP FP'='SNP_FP',
  'SNP FN'='SNP_FN',
  'INDEL truth'='IND_Truth',
  'INDEL TP'='IND_TP',
  'INDEL FP'='IND_FP',
  'IND FN'='IND_FN'
)
# Rounded columns
v.smlv_summary_table_columns_round <- c(
  'SNP recall',
  'SNP precision',
  'INDEL recall',
  'INDEL precision'
)
# Columns formatted with commas
v.smlv_summary_table_columns_comma <- c(
  'Run 1 count',
  'Run 2 count',
  'SNP truth',
  'SNP TP',
  'SNP FP',
  'SNP FN',
  'INDEL truth',
  'INDEL TP',
  'INDEL FP',
  'IND FN'
)
```

```{r smlv_summary_data}
get_smlv_summary_data <- function(s.base_dir) {
  # Read in counts
  s.count_fp <- paste0(s.base_dir, '/small_variants/counts.tsv')
  if (! fs::file_exists(s.count_fp)) {
    return(NULL)
  }
  d.counts <- readr::read_tsv(s.count_fp, col_types=readr::cols())
  # Select and read in all comparison files
  v.comp_fps <- Sys.glob(paste0(s.base_dir, '/small_variants/3_comparison/*tsv'))
  v.comp <- lapply(v.comp_fps, readr::read_tsv, col_types=readr::cols())

  # Process counts data
  # 1. remove intersect VCFS - identified by run == 'n/a'
  # 2. recode values: source: input -> all; run: first -> run1_count, second -> run2_count
  # 3. cast to wider format - remove filename col; add first and second columns
  d.counts_wide <- d.counts %>%
    dplyr::filter(run!='n/a') %>%
    dplyr::mutate(
      source=dplyr::recode(source, input='all'),
      run=dplyr::recode(run, first='run1_count', second='run2_count'),
    ) %>%
    tidyr::pivot_wider(names_from=run, id_cols=c(vcf_type, source), values_from=count)

  # Create small variants table
  # 1. bind individual comparison tables into one
  # 2. full join on comparison ~ counts; i.e. add counts to comparison table
  # 3. select, order, and rename desired columns
  # 4. round precision/recall and add commas to other columns
  d.smlv_table <- dplyr::bind_rows(v.comp) %>%
    dplyr::full_join(d.counts_wide, by=c('flabel'='vcf_type', 'subset'='source')) %>%
    dplyr::select(dplyr::all_of(v.smlv_summary_table_columns)) %>%
    dplyr::mutate(dplyr::across(dplyr::all_of(v.smlv_summary_table_columns_round), function(v.col) { format(round(v.col, 4), nsmall=4) })) %>%
    dplyr::mutate(dplyr::across(dplyr::all_of(v.smlv_summary_table_columns_comma), scales::comma))
  return(d.smlv_table)
}

# Finalise small variant data
# 1. bind rows into single table
# 2. sort rows by subset
d.smlv_summary_data <- purrr::map(v.sample_dirs, get_smlv_summary_data) %>%
  dplyr::bind_rows() %>%
  dplyr::arrange(Subset)
```

```{r smlv_summary_table}
n.table_height <- min(38.75 * nrow(d.smlv_summary_data), 600)
d.smlv_summary_data %>%
  DT::datatable(
    rownames=FALSE,
    filter=list(position='top', clear=FALSE, plain=TRUE),
    extensions='Scroller',
    options=list(scroller=TRUE, scrollX=TRUE, scrollY=n.table_height)
  ) %>%
  DT::formatStyle(
    v.smlv_summary_table_columns_round,
    background=DT::styleColorBar(as.numeric(unlist(d.smlv_summary_data[ ,v.smlv_summary_table_columns_round])), 'lightgreen'),
    backgroundSize='90% 90%',
    backgroundRepeat='no-repeat',
    backgroundPosition='center'
  )
```

## PCGR tables
```{r smlv_pcgr_constants}
# Small variant PCGR table columns
# Order and display name
v.smlv_vcf_columns <- c(
  'Sample name'='sample_name',
  'Chromosome'='CHROM',
  'Position'='POS',
  'PCGR tier'='PCGR_TIER',
  'PCGR symbol'='PCGR_SYMBOL',
  'Quality score'='QUAL',
  'Tumor AF'='TUMOR_AF',
  'Normal AF'='NORMAL_AF',
  'Callers'='CALLERS',
  'SAGE hotspot'='SAGE_HOTSPOT'
)
# bcftools isec filenames and call types
v.files <- c(
  'fp'= '0000.vcf',
  'fn'='0001.vcf'
)
# PCGR variant type display order and names
v.pcgr_variant_types <- c(
  'Tier 1'='TIER_1',
  'Tier 2'='TIER_2',
  'Tier 3'='TIER_3',
  'Tier 4'='TIER_4',
  'Non-coding'='NONCODING'
)
```

```{r smlv_pcgr_data}
get_smlv_pcgr_data <- function(s.base_dir) {
  s.sample_name <- basename(s.base_dir)
  v.pcgr_vcfs <- purrr::map(names(v.files), function(s.vcf_type) {
    # Collect variant data
    # 1. read in vcf, select data section, and convert to tibble
    # 2. select specific rows, if they are present (using `values` to delay renaming columns until later)
    # 3. add column for (1) sample name, and (2) to identify variant as fp or fn
    # 4. add missing columns
    s.vcf_filename <- v.files[s.vcf_type]
    s.vcf_fp <- paste0(s.base_dir, '/small_variants/2_variants_intersect/input/pcgr/', s.vcf_filename)
    if (! fs::file_exists(s.vcf_fp)) {
      return(NULL)
    }
    d.vcf <- bedr::read.vcf(s.vcf_fp, split.info=TRUE, verbose=FALSE) %>%
      purrr::pluck('vcf') %>%
      tibble::as_tibble() %>%
      dplyr::select(unname(v.smlv_vcf_columns)[v.smlv_vcf_columns %in% colnames(.)]) %>%
      dplyr::mutate(sample_name=s.sample_name, call_errortype=s.vcf_type)
    # Add any missing columns
    v.columns_missing <- v.smlv_vcf_columns[! v.smlv_vcf_columns %in% colnames(d.vcf)]
    v.columns_missing <- v.columns_missing[v.columns_missing!='sample_name']
    d.vcf[ ,v.columns_missing] <- NA
    # Rename and order columns
    d.vcf <- dplyr::select(d.vcf, dplyr::all_of(v.smlv_vcf_columns))
    # Return NULL if vcf is empty
    if (nrow(d.vcf)==0) {
      return(NULL)
    } else {
      return(d.vcf)
    }
  })
  # If both vcfs contain no data, return NULL
  if (all(sapply(v.pcgr_vcfs, is.null))) {
    return(NULL)
  }
  # Otherwise, finalise variant data
  # 1. merge fp and fn into single tibble
  # 2. order entries by chromosome number
  # 3. place the 'sample_name' column first
  # 4. format position with commas
  dplyr::bind_rows(v.pcgr_vcfs) %>%
    dplyr::arrange(Chromosome) %>%
    dplyr::select('Sample name', dplyr::everything()) %>%
    dplyr::mutate(Position=scales::comma(Position, accuracy=1))
}

# Finalise PCGR data
# 1. bind rows into single table
# 2. sort rows by subset
d.smlv_pcgr_data <- purrr::map(v.sample_dirs, get_smlv_pcgr_data) %>%
  dplyr::bind_rows()

# Subset top n, ordering by supposed priority
#   1. group by sample and then order by PCGR tier
#   2. select top n entries
d.smlv_pcgr_data_subset <- d.smlv_pcgr_data %>%
  dplyr::group_by(`Sample name`) %>%
  dplyr::arrange(desc(`SAGE hotspot`), desc(`PCGR tier`)) %>%
  dplyr::slice(1:10)
```

```{r smlv_pcgr_tablulate_function}
# Set missing PCGR categories; required for selecting and ordering columns
v.missing <- v.pcgr_variant_types[! v.pcgr_variant_types %in% unique(d.smlv_pcgr_data$`PCGR tier`)]

tabulate_pcgr_tiers <- function(s.var) {
  # Create tabulated counts
  # 1. group data by PCGR tier and given variable
  # 2. tally counts
  # 3. convert from grouped tbl back to ungrouped
  # 3. cast into wide format
  d.table_counts <- d.smlv_pcgr_data %>%
    dplyr::group_by(!!dplyr::sym(s.var), `PCGR tier`) %>%
    dplyr::count(.drop=FALSE) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_wider(names_from=`PCGR tier`, id_cols=!!dplyr::sym(s.var), values_from=n, values_fill=0)
  # Complete processing of tabulated data
  # 1. add missing rows
  # 2. order and rename with grouping variable in first position and unknown categories at end
  # 3. add botton sum margin
  # 4. set rowwise, and then compute and add right sum margin
  # 5. convert from rowwise tbl to ungrouped
  d.table_counts[ ,v.missing] <- 0
  dplyr::select(d.table_counts, c(!!dplyr::sym(s.var), dplyr::all_of(v.pcgr_variant_types)), dplyr::everything()) %>%
    dplyr::bind_rows(
      dplyr::summarise(
        .,
        dplyr::across(where(is.numeric), sum),
        dplyr::across(where(is.character), ~'Total'))
    ) %>%
    dplyr::rowwise(.) %>%
    dplyr::mutate(Total=sum(dplyr::c_across(where(is.numeric)))) %>%
    dplyr::ungroup()
}
```

### Variants counts by sample
```{r smlv_pcgr_table_sample_name}
# Tabulate by sample name
tabulate_pcgr_tiers('Sample name') %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width=FALSE, position='left')
```

### Variants counts by SAGE hotspot
```{r smlv_pcgr_table_sage}
# Tabulate by sage hotspot
tabulate_pcgr_tiers('SAGE hotspot') %>%
  knitr::kable() %>%
  kableExtra::kable_styling(full_width=FALSE, position='left')
```

### Variants
```{r smlv_pcgr_table_all_variants}
# Subset of individual PCGR variants
n.table_height <- min(38.75 * nrow(d.smlv_pcgr_data_subset), 600)
d.smlv_pcgr_data_subset %>%
  DT::datatable(
    rownames=FALSE,
    caption='PCGR FP and FN variants for all samples.',
    filter=list(position='top',clear=FALSE, plain=TRUE),
    extensions=c('Scroller', 'Buttons'),
    options=list(scroller=TRUE, scrollX=TRUE, scrollY=n.table_height, dom='Bfrtip', buttons=c('csv'))
  )
```
