# Copy number variants {.tabset .tabset-fade}
## Summary
```{r cnv_summary_constants}
v.cnv_table_columns <- c(
  'Sample name'='sample_name',
  'Gene'='gene',
  'UMCCR gene'='in_umccr',
  'Different lower estimate'='min_diff',
  'Different upper estimate'='max_diff',
  'Lower estimate (first/second)'='mincn_1_2',
  'Upper estimate (first/second)'='maxcn_1_2',
  'Lower estimate difference'='mincn_diff',
  'Upper estimate difference'='maxcn_diff',
  'Chromosome'='chrom',
  'Start'='start',
  'End'='end'
)
v.cnv_table_columns_comma <- c(
  'Start',
  'End'
)
v.cnv_table_columns_colour <- c(
  'UMCCR gene',
  'Different lower estimate',
  'Different upper estimate'
)
```

```{r cnv_summary_functions}
discover_cnv_files <- function(s.base_dir) {
  # Get sample and and run type from dirpath
  s.sample_name <- fs::path_dir(s.base_dir) %>% basename()
  s.run_type <- basename(s.base_dir)
  # NOTE: we do not expect any files for only bcbio
  if (s.run_type == 'bcbio') {
    return(create_empty_file_info())
  }
  s.cnv_fp <- s.base_dir / '/copy_number_variants/cn_diff.tsv'
  tibble::tibble(
    sample_name=s.sample_name,
    file_type='cnv_diff',
    file=s.cnv_fp,
    exists=fs::file_exists(s.cnv_fp)
  )
}

get_cnv_summary_data <- function(d.files, v.umccr_genes) {
  # Prepare copy number variant data
  # 1. read CN data files, and add sample name column
  # 2. create new columns:
  #      - combine min/max for run 1 and run 2,
  #      - difference of min/max b/n runs, and
  #      - is a UMCCR gene
  # 3. recode TRUE/FALSE as Yes/No
  # 4. order and rename columns
  # 5. add format appropriate numeric columns with commas
  d.cnv_file <- d.files[d.files$file_type=='cnv_diff', ]
  s.cnv_file <- d.cnv_file$file
  s.sample_name <- d.cnv_file$sample_name
  d.cnv_data_diff <- readr::read_tsv(s.cnv_file, col_types='ciicddddcc') %>%
    dplyr::mutate(sample_name=s.sample_name) %>%
    dplyr::mutate(
      mincn_1_2=paste0(min_cn.run1, '/', min_cn.run2),
      maxcn_1_2=paste0(max_cn.run1, '/', max_cn.run2),
      mincn_diff=round(abs(min_cn.run1 - min_cn.run2), 2),
      maxcn_diff=round(abs(max_cn.run1 - max_cn.run2), 2),
      in_umccr=ifelse(gene %in% v.umccr_genes, 'Yes', 'No') %>% as.character()
    ) %>%
    dplyr::mutate(
      min_diff=dplyr::recode(min_diff, 'TRUE'='Yes', 'FALSE'='No'),
      max_diff=dplyr::recode(max_diff, 'TRUE'='Yes', 'FALSE'='No')
    ) %>%
    dplyr::select(dplyr::all_of(v.cnv_table_columns)) %>%
    dplyr::mutate(across(dplyr::all_of(v.cnv_table_columns_comma), scales::comma))
}

render_cnv_summary_table <- function(d.cnv_summary_data_subset) {
  n.table_height <- min(38.75 * nrow(d.cnv_summary_data_subset), 600)
  d.cnv_summary_data_subset %>%
    dplyr::mutate_if(is.character, as.factor) %>%
    DT::datatable(
      rownames=FALSE,
      caption=paste0('Copy number bound difference threshold: 0.1'),
      filter=v.table_filter_options,
      extensions=c('Scroller', 'KeyTable'),
      options=list(
        scroller=TRUE,
        scrollX=TRUE,
        scrollY=n.table_height
      )
    ) %>%
    DT::formatStyle(
      v.cnv_table_columns_colour,
      backgroundColor=DT::styleEqual(
        'Yes',
        v.table_cell_fill_colour_red
      )
    )
}
```

```{r cnv_summary_process}
# NOTE: may be possible to generalise cleanly without too much work
v.cnv_files <- purrr::map(v.run_dirs, discover_cnv_files)
v.cnv_files_exist <- v.cnv_files %>% purrr::keep(function(d) { any(d$exists) })
# Process data or create empty tibble
if (length(v.cnv_files_exist) > 0) {
  # Read in some necessary data
  s.umcccr_genes_fp <- system.file('extdata/genes/umccr_cancer_genes.latest.genes', package='woofr')
  v.umccr_genes <- readr::read_lines(s.umcccr_genes_fp)
  # Get and finalise CNV data
  d.cnv_summary_data <- purrr::map(v.cnv_files_exist, get_cnv_summary_data, v.umccr_genes) %>%
    dplyr::bind_rows()
  # Subset top n, ordering by supposed priority
  d.cnv_summary_data_sorted <- d.cnv_summary_data %>%
    dplyr::arrange(-(`Lower estimate difference` + `Upper estimate difference`))
  d.cnv_summary_data_subset <- subset_variants(d.cnv_summary_data_sorted, v.table_display_max)
} else {
  d.cnv_summary_data <- d.cnv_summary_data_subset <- dplyr::bind_rows(v.cnv_table_columns)[NULL, ]
}
```

```{r cnv_warnings}
# NOTE: this pattern and the one below could probably be generalised as a fn
# Missing file warning
d.cnv_files <- dplyr::bind_rows(v.cnv_files)
if (nrow(d.cnv_files) >= 1) {
  d.cnv_files_missing <- d.cnv_files[! d.cnv_files$exists, ]
  l.cnv_files_missing <- nrow(d.cnv_files_missing) > 0
} else {
  l.cnv_files_missing <- FALSE
}

# Undisplayed variant warning
if (nrow(d.cnv_summary_data_subset) > 0) {
  # Set flag
  d.cnv_summary_data_displayed <- get_variant_display_data(d.cnv_summary_data, d.cnv_summary_data_subset)
  l.cnv_undisplayed_data <- any(d.cnv_summary_data_displayed$`Not displayed` > 0)
} else {
  l.cnv_undisplayed_data <- FALSE
}
```

```{r cnv_summary_data_warning_files, eval=l.cnv_files_missing, results='asis'}
render_warning_files(d.cnv_files_missing)
```

```{r cnv_summary_data_warning_display, eval=l.cnv_undisplayed_data, results='asis'}
render_warning_variants(d.cnv_summary_data_displayed)
```

```{r cnv_summary_table}
render_cnv_summary_table(d.cnv_summary_data_subset)
```
