# Copy number variants {.tabset .tabset-fade}
```{r cnv_source_helpers}
source('warning_functions.R')
```

## Summary
```{r cnv_summary_constants}
v.cnv_table_columns <- c(
  'Sample name'='sample_name',
  'Gene'='gene',
  'UMCCR gene'='in_umccr',
  'Min different'='min_diff',
  'Max different'='max_diff',
  'Min (first/second)'='mincn_1_2',
  'Max (first/second)'='maxcn_1_2',
  'Min diff'='mincn_diff',
  'Max diff'='maxcn_diff',
  'Chromosome'='chrom',
  'Start'='start',
  'End'='end'
)
v.cnv_table_columns_comma <- c(
  'Start',
  'End'
)
v.cnv_table_columns_colour <- c(
  'UMCCR gene',
  'Min different',
  'Max different'
)
```

```{r cnv_summary_functions}
discover_cnv_files <- function(s.base_dir) {
  s.cnv_fp <- s.base_dir / '/copy_number_variants/cn_diff.tsv'
  tibble::tibble(
    sample_name=basename(s.base_dir),
    file_type='cnv_diff',
    file=s.cnv_fp,
    exists=fs::file_exists(s.cnv_fp)
  )
}

get_cnv_summary_data <- function(d.files, v.umccr_genes) {
  # Prepare copy number variant data
  # 1. read CN data files, and add sample name column
  # 2. create new columns:
  #      - combine min/max for run 1 and run 2,
  #      - difference of min/max b/n runs, and
  #      - is a UMCCR gene
  # 3. order and rename colums
  # 4. add format appropriate numeric columns with commas
  d.cnv_file <- d.files[d.files$file_type=='cnv_diff', ]
  s.cnv_file <- d.cnv_file$file
  s.sample_name <- d.cnv_file$sample_name
  d.cnv_data_diff <- readr::read_tsv(s.cnv_file, col_types='ciicddddcc') %>%
    dplyr::mutate(sample_name=s.sample_name) %>%
    dplyr::mutate(
        mincn_1_2=paste0(min_cn.run1, '/', min_cn.run2),
        maxcn_1_2=paste0(max_cn.run1, '/', max_cn.run2),
        mincn_diff=round(abs(min_cn.run1 - min_cn.run2), 2),
        maxcn_diff=round(abs(max_cn.run1 - max_cn.run2), 2),
        in_umccr=ifelse(gene %in% v.umccr_genes, 'TRUE', 'FALSE')
    ) %>%
    dplyr::select(dplyr::all_of(v.cnv_table_columns)) %>%
    dplyr::mutate(across(dplyr::all_of(v.cnv_table_columns_comma), scales::comma))
}

render_cnv_summary_table <- function(d.cnv_summary_data_subset) {
  n.table_height <- min(31 * nrow(d.cnv_summary_data_subset), 600)
  d.cnv_summary_data_subset %>%
    dplyr::mutate_if(is.character, as.factor) %>%
    DT::datatable(
      rownames=FALSE,
      class='cell-border display compact',
      caption=paste0('Differences in Copy Number (threshold: 0.1)'),
      filter=list(position='top', clear=FALSE, plain=FALSE),
      extensions=c('Scroller', 'KeyTable'),
      options=list(
        scroller=TRUE,
        scrollX=TRUE,
        scrollY=n.table_height
      )
    ) %>%
    DT::formatStyle(
      v.cnv_table_columns_colour,
      backgroundColor=DT::styleEqual(c('TRUE'), c('#FFCCCB'))
    )
}
```

```{r cnv_summary_process}
# NOTE: may be possible to generalise cleanly without too much work
v.cnv_files <- purrr::map(v.sample_dirs, discover_cnv_files)
v.cnv_files_exist <- v.cnv_files[sapply(v.cnv_files, function(d) { d$exists })]
# Process data or create empty tibble
if (length(v.cnv_files_exist) > 0) {
  # Read in some necessary data
  s.umcccr_genes_fp <- system.file('extdata/genes/umccr_cancer_genes.latest.genes', package='woofr')
  v.umccr_genes <- readr::read_lines(s.umcccr_genes_fp)
  # Get and finalise CNV data
  d.cnv_summary_data <- purrr::map(v.cnv_files, get_cnv_summary_data, v.umccr_genes) %>%
    dplyr::bind_rows()
  # Subset data for display
  #  1. order by size of min/max difference
  #  2. select up to n entries
  d.cnv_summary_data_subset <- d.cnv_summary_data %>%
    dplyr::arrange(-(`Max diff` + `Max diff`)) %>%
    dplyr::slice(1:10)
} else {
  d.cnv_summary_data <- d.cnv_summary_data_subset <- dplyr::bind_rows(v.cnv_table_columns)[NULL, ]
}
```

```{r cnv_warnings}
# NOTE: this pattern and the one below could probably be generalised as a fn
# Missing file warning
d.cnv_files <- dplyr::bind_rows(v.cnv_files)
d.cnv_files_missing <- d.cnv_files[! d.cnv_files$exists, ]
l.cnv_files_missing <- nrow(d.cnv_files_missing) > 0

# Undisplayed variant warning
if (nrow(d.cnv_summary_data_subset) > 0) {
  # Set flag
  d.cnv_summary_data_displayed <- get_variant_display_data(d.cnv_summary_data, d.cnv_summary_data_subset)
  l.cnv_undisplayed_data <- any(d.cnv_summary_data_displayed$`Not displayed` > 0)
} else {
  l.cnv_undisplayed_data <- FALSE
}
```

```{r cnv_summary_data_warning_files, eval=l.cnv_files_missing, results='asis'}
render_warning_files(d.cnv_files_missing)
```

```{r cnv_summary_data_warning_display, eval=l.cnv_undisplayed_data, results='asis'}
render_warning_variants(d.cnv_summary_data_displayed)
```

```{r cnv_summary_table}
render_cnv_summary_table(d.cnv_summary_data_subset)
```
