# Small variants {.tabset .tabset-fade}
## Summary
```{r smlv_summary_constants}
# Small variant summary table columns
# Order and display name
v.smlv_summary_table_columns <- c(
  'Sample name'='sample',
  'Run type'='run_type',
  'VCF source'='flabel',
  'Subset'='subset',
  'Run 1 count'='run1_count',
  'Run 2 count'='run2_count',
  'SNP recall'='SNP_Recall',
  'SNP precision'='SNP_Precision',
  'INDEL recall'='IND_Recall',
  'INDEL precision'='IND_Precision',
  'SNP truth'='SNP_Truth',
  'SNP TP'='SNP_TP',
  'SNP FP'='SNP_FP',
  'SNP FN'='SNP_FN',
  'INDEL truth'='IND_Truth',
  'INDEL TP'='IND_TP',
  'INDEL FP'='IND_FP',
  'IND FN'='IND_FN'
)
# Rounded columns
v.smlv_summary_table_columns_round <- c(
  'SNP recall',
  'SNP precision',
  'INDEL recall',
  'INDEL precision'
)
# Columns formatted with commas
v.smlv_summary_table_columns_comma <- c(
  'Run 1 count',
  'Run 2 count',
  'SNP truth',
  'SNP TP',
  'SNP FP',
  'SNP FN',
  'INDEL truth',
  'INDEL TP',
  'INDEL FP',
  'IND FN'
)
# Expected comparison files
v.comparison_file_types <- list(
  'umccrise'=c(
    'pcgr',
    'pcgr_filtered',
    'cpsr',
    'cpsr_filtered'
  ),
  'bcbio'=c(
    'ensemble',
    'ensemble_filtered',
    'gatk',
    'gatk_filtered',
    'strelka2',
    'strelka2_filtered',
    'vardict',
    'vardict_filtered'
  )
)
```

```{r smlv_summary_functions}
discover_smlv_files <- function(s.base_dir) {
  # Get sample and and run type from dirpath
  s.sample_name <- fs::path_dir(s.base_dir) %>% basename()
  s.run_type <- basename(s.base_dir)
  # Variant counts
  s.count_fp <- s.base_dir / '/small_variants/counts.tsv'
  d.file_info_counts <- tibble::tibble(
    sample_name=s.sample_name,
    run_type=s.run_type,
    file_type='count',
    file=s.count_fp,
    exists=fs::file_exists(s.count_fp)
  )
  # Comparison files
  v.file_info_comparisons <- purrr::map(v.comparison_file_types[[s.run_type]], function(s.file_type) {
    s.fp <- s.base_dir / 'small_variants/3_comparison' / paste0(s.file_type, '.tsv')
    tibble::tibble(
      sample_name=s.sample_name,
      run_type=s.run_type,
      file_type=s.file_type,
      file=s.fp,
      exists=fs::file_exists(s.fp)
    )
  })
  d.file_info_comparisons <- dplyr::bind_rows(v.file_info_comparisons)
  # Combine, set whether count info is available, and return
  d.files <- dplyr::bind_rows(
    d.file_info_counts,
    d.file_info_comparisons
  )
  d.files$has_counts <- d.files$exists[d.files$file_type=='count'] && any(d.files$exists[d.files$file_type %in% v.comparison_file_types[[s.run_type]]])
  return(d.files)
}

get_smlv_summary_data <- function(d.files) {
  # Process counts data
  # 1. read in count data
  # 2. remove intersect VCFS - identified by run == 'n/a'
  # 3. recode values: source: input -> all; run: one -> run1_count, two -> run2_count
  # 4. cast to wider format - remove filename col; add `one` and `two` columns
  s.count_file <- d.files$file[d.files$file_type=='count']
  d.counts_wide <- readr::read_tsv(s.count_file, col_types=readr::cols()) %>%
    dplyr::filter(run!='n/a') %>%
    dplyr::mutate(
      source=dplyr::recode(source, input='all'),
      run=dplyr::recode(run, one='run1_count', two='run2_count'),
    ) %>%
    tidyr::pivot_wider(names_from=run, id_cols=c(vcf_type, source), values_from=count)

  # Create small variants table
  # 1. bind individual comparison tables into one
  # 2. full join on comparison ~ counts; i.e. add counts to comparison table
  # 3. add run type
  # 4. select, order, and rename desired columns
  # 5. round precision/recall and add commas to other columns
  s.run_type <- unique(d.files$run_type)
  d.comparison_files_all <- d.files[d.files$file_type %in% v.comparison_file_types[[s.run_type]], ]
  v.comparison_files <- d.comparison_files_all$file[d.comparison_files_all$exists]
  d.smlv_table <- purrr::map(v.comparison_files, readr::read_tsv, col_types=readr::cols()) %>%
    dplyr::bind_rows() %>%
    dplyr::left_join(d.counts_wide, by=c('flabel'='vcf_type', 'subset'='source')) %>%
    dplyr::mutate(run_type=s.run_type) %>%
    dplyr::select(dplyr::all_of(v.smlv_summary_table_columns)) %>%
    dplyr::mutate(dplyr::across(
      dplyr::all_of(v.smlv_summary_table_columns_round),
      function(v.col) { format(round(v.col, 4), nsmall=4, scientific=FALSE) })
    ) %>%
    dplyr::mutate(dplyr::across(dplyr::all_of(v.smlv_summary_table_columns_comma), scales::comma))
  return(d.smlv_table)
}

render_smlv_summary_table <- function(d.smlv_summary_data) {
  n.table_height <- min(38.75 * nrow(d.smlv_summary_data), 600)
  d.smlv_summary_data %>%
    DT::datatable(
      rownames=FALSE,
      filter=v.table_filter_options,
      extensions=c('Scroller', 'KeyTable'),
      options=list(
        scroller=TRUE,
        scrollX=TRUE,
        scrollY=n.table_height
      )
    ) %>%
    DT::formatStyle(
      v.smlv_summary_table_columns_round,
      background=DT::styleColorBar(
        v.table_cell_fill_range,
        v.table_cell_fill_colour_green
      )
    )
}
```

```{r smlv_summary_process}
# NOTE: we should consider whether we need to subset variants for display
# NOTE: may be possible to generalise cleanly without too much work
v.smlv_files <- purrr::map(v.run_dirs, discover_smlv_files)
v.smlv_files_with_counts <- v.smlv_files %>% purrr::keep(function(d) { any(d$exists) })
# Process data or create empty tibble
if (length(v.smlv_files_with_counts) > 0) {
  # Get and finalise small variant data
  # 1. bind rows into single table
  # 2. sort rows by subset
  d.smlv_summary_data <- purrr::map(v.smlv_files, get_smlv_summary_data) %>%
    dplyr::bind_rows() %>%
    dplyr::arrange(Subset)
} else {
  d.smlv_summary_data <- dplyr::bind_rows(v.smlv_summary_table_columns)[NULL, ]
}
```

```{r smlv_summary_warnings}
d.smlv_files <- dplyr::bind_rows(v.smlv_files)
if (nrow(d.smlv_files) >= 1) {
  d.smlv_files_missing <- d.smlv_files[! d.smlv_files$exists, ]
  l.smlv_files_missing <- nrow(d.smlv_files_missing) > 0
} else {
  l.smlv_files_missing <- FALSE
}

# NOTE: if we decide to subset variants for display, warning logic will be needed here and then rendered below
```

```{r smlv_summary_render_warning_files, eval=l.smlv_files_missing, results='asis'}
render_warning_files(d.smlv_files_missing)
```

```{r smlv_summary_table}
# Render
render_smlv_summary_table(d.smlv_summary_data)
```

## PCGR
```{r smlv_pcgr_constants}
# Small variant PCGR table columns
# Order and display name
v.smlv_vcf_columns <- c(
  'Sample name'='sample_name',
  'Chromosome'='CHROM',
  'Position'='POS',
  'PCGR tier'='PCGR_TIER',
  'PCGR symbol'='PCGR_SYMBOL',
  'Quality score'='QUAL',
  'Tumor AF'='TUMOR_AF',
  'Normal AF'='NORMAL_AF',
  'Callers'='CALLERS',
  'SAGE hotspot'='SAGE_HOTSPOT'
)
# bcftools isec filenames and call types
v.files <- c(
  'fp'= '0000.vcf',
  'fn'='0001.vcf'
)
# PCGR variant type display order and names
v.pcgr_variant_types <- c(
  'Tier 1'='TIER_1',
  'Tier 2'='TIER_2',
  'Tier 3'='TIER_3',
  'Tier 4'='TIER_4',
  'Non-coding'='NONCODING'
)
```

```{r smlv_pcgr_functions}
discover_smlv_pcgr_files <- function(s.base_dir) {
  # Get sample and and run type from dirpath
  s.sample_name <- fs::path_dir(s.base_dir) %>% basename()
  s.run_type <- basename(s.base_dir)
  # NOTE: we do not expect any files for only bcbio
  if (s.run_type == 'bcbio') {
    return(create_empty_file_info())
  }
  v.pcgr_vcfs <- purrr::map(names(v.files), function(s.vcf_type) {
    s.vcf_filename <- v.files[s.vcf_type]
    s.vcf_fp <- s.base_dir / 'small_variants/2_variants_intersect/input/pcgr' / s.vcf_filename
    tibble::tibble(
      sample_name=s.sample_name,
      run_type=s.run_type,
      file_type=s.vcf_type,
      file=s.vcf_fp,
      exists=fs::file_exists(s.vcf_fp)
    )
  })
}

get_smlv_pcgr_data <- function(d.file) {
  # Collect variant data
  # 1. read in vcf, select data section, and convert to tibble
  # 2. select specific rows, if they are present (using `values` to delay renaming columns until later)
  # 3. add column for (1) sample name, and (2) to identify variant as fp or fn
  # 4. add missing columns
  d.vcf <- bedr::read.vcf(d.file$file, split.info=TRUE, verbose=FALSE) %>%
    purrr::pluck('vcf') %>%
    tibble::as_tibble() %>%
    dplyr::select(unname(v.smlv_vcf_columns)[v.smlv_vcf_columns %in% colnames(.)]) %>%
    dplyr::mutate(sample_name=d.file$sample_name, call_errortype=d.file$file_type)
  # Add any missing columns
  v.columns_missing <- v.smlv_vcf_columns[! v.smlv_vcf_columns %in% colnames(d.vcf)]
  v.columns_missing <- v.columns_missing[v.columns_missing!='sample_name']
  d.vcf[ ,v.columns_missing] <- NA
  # Rename and order columns
  d.vcf <- dplyr::select(d.vcf, dplyr::all_of(v.smlv_vcf_columns))
  # Return NULL if vcf is empty
  if (nrow(d.vcf)==0) {
    return(NULL)
  } else {
    return(d.vcf)
  }
}

tabulate_pcgr_tiers <- function(d.smlv_pcgr_data, s.var) {
  # Set missing entries
  v.missing <- v.pcgr_variant_types[! v.pcgr_variant_types %in% unique(d.smlv_pcgr_data$`PCGR tier`)]
  # Create tabulated counts
  # 1. group data by PCGR tier and given variable
  # 2. tally counts
  # 3. convert from grouped tbl back to ungrouped
  # 3. cast into wide format
  d.table_counts <- d.smlv_pcgr_data %>%
    dplyr::group_by(!!dplyr::sym(s.var), `PCGR tier`) %>%
    dplyr::count(.drop=FALSE) %>%
    dplyr::ungroup() %>%
    tidyr::pivot_wider(names_from=`PCGR tier`, id_cols=!!dplyr::sym(s.var), values_from=n, values_fill=0)
  # Complete processing of tabulated data
  # 1. add missing rows
  # 2. order and rename with grouping variable in first position and unknown categories at end
  # 3. add bottom sum margin
  # 4. set rowwise, and then compute and add right sum margin
  # 5. convert from rowwise tbl to ungrouped
  d.table_counts[ ,v.missing] <- 0
  dplyr::select(d.table_counts, c(!!dplyr::sym(s.var), dplyr::all_of(v.pcgr_variant_types)), dplyr::everything()) %>%
    dplyr::bind_rows(
      dplyr::summarise(
        .,
        dplyr::across(where(is.numeric), sum),
        dplyr::across(where(is.character), ~'Total'))
    ) %>%
    dplyr::rowwise(.) %>%
    dplyr::mutate(Total=sum(dplyr::c_across(where(is.numeric)))) %>%
    dplyr::ungroup()
}

smlv_pcgr_render_cross_table <- function(d.smlv_pcgr_data, s.variable) {
  tabulate_pcgr_tiers(d.smlv_pcgr_data, s.variable) %>%
    knitr::kable() %>%
    kableExtra::kable_styling(full_width=FALSE, position='left')
}

smlv_pcgr_render_table <- function(d.smlv_pcgr_data_subset) {
  n.table_height <- min(38.75 * nrow(d.smlv_pcgr_data_subset), 600)
  d.smlv_pcgr_data_subset %>%
    DT::datatable(
      rownames=FALSE,
      caption='PCGR FP and FN variants for all samples.',
      filter=v.table_filter_options,
      extensions=c('Scroller', 'KeyTable'),
      options=list(
        scroller=TRUE,
        scrollX=TRUE,
        scrollY=n.table_height
      )
    )
}
```

```{r smlv_pcgr_process}
v.smlv_pcgr_files_nested <- purrr::map(v.run_dirs, discover_smlv_pcgr_files)
# Here I attempt to flatten the sometimes nested list of file info tibbles. We can get either
# and empty tibble or a list of two tibbles. Not immediately obvious if there is a clean tidyverse
# approach to this.
# NOTE: as a side effect of c(), empty tibbles are removed.
v.smlv_pcgr_files <- NULL
for (v.set in v.smlv_pcgr_files_nested) {
  if ('tbl_df' %in% class(v.set)) {
    v.smlv_pcgr_files <- c(v.smlv_pcgr_files, purrr::flatten(v.set))
  } else {
    v.smlv_pcgr_files <- c(v.smlv_pcgr_files, v.set)
  }
}
v.smlv_pcgr_files_exist <- v.smlv_pcgr_files %>% purrr::keep(function(d) { any(d$exists) })
# Process data or create empty tibble
if (length(v.smlv_pcgr_files_exist) > 0) {
  # Get and finalise PCGR data
  # 1. bind rows into single table
  # 2. format position column with comma
  d.smlv_pcgr_data <- purrr::map(v.smlv_pcgr_files_exist, get_smlv_pcgr_data) %>%
    dplyr::bind_rows() %>%
    dplyr::mutate(Position=scales::comma(Position, accuracy=1))
  # Subset top n, ordering by supposed priority
  d.smlv_pcgr_data_sorted <- d.smlv_pcgr_data %>%
    dplyr::arrange(desc(`SAGE hotspot`), desc(`PCGR tier`))
  d.smlv_pcgr_data_subset <- subset_variants(d.smlv_pcgr_data_sorted, v.table_display_max)
} else {
  d.smlv_pcgr_data <- d.smlv_pcgr_data_subset <- dplyr::bind_rows(v.smlv_vcf_columns)[NULL, ]
}
```

```{r smlv_pcgr_warnings}
# NOTE: this pattern and the one below could probably be generalised as a fn
# Missing file warning
d.smlv_pcgr_files <- dplyr::bind_rows(v.smlv_pcgr_files)
if (nrow(d.smlv_pcgr_files) >= 1) {
  d.smlv_pcgr_files_missing <- d.smlv_pcgr_files[! d.smlv_pcgr_files$exists, ]
  l.smlv_pcgr_files_missing <- nrow(d.smlv_pcgr_files_missing) > 0
} else {
  l.smlv_pcgr_files_missing <- FALSE
}

# Undisplayed variant warning
if (nrow(d.smlv_pcgr_data_subset) > 0) {
  # Set flag
  d.smlv_pcgr_data_displayed <- get_variant_display_data(d.smlv_pcgr_data, d.smlv_pcgr_data_subset)
  l.undisplayed_smlv_pcgr_variants <- any(d.smlv_pcgr_data_displayed$`Not displayed` > 0)
} else {
  l.undisplayed_smlv_pcgr_variants <- FALSE
}
```

```{r smlv_pcgr_variants_warning_files, eval=l.smlv_pcgr_files_missing, results='asis'}
render_warning_files(d.smlv_pcgr_files_missing)
```

### Variants counts by sample
```{r smlv_pcgr_table_sample_name}
smlv_pcgr_render_cross_table(d.smlv_pcgr_data, 'Sample name')
```

### Variants counts by SAGE hotspot
```{r smlv_pcgr_table_sage}
smlv_pcgr_render_cross_table(d.smlv_pcgr_data, 'SAGE hotspot')
```

### Variants
```{r smlv_pcgr_variants_warning_display, eval=l.undisplayed_smlv_pcgr_variants, results='asis'}
render_warning_variants(d.smlv_pcgr_data_displayed)
```

```{r smlv_pcgr_table_all_variants}
smlv_pcgr_render_table(d.smlv_pcgr_data_subset)
```
